Lhk 551 change printing from console to log

olegivchenkolula opened the pull request:
Added logging to HttpHandler

olegivchenkolula committed:
Added Playground and voyager for LHK-542

olegivchenkolula committed:
Fixed introspection returning

olegivchenkolula committed:
Removed redundant open

olegivchenkolula committed:
Code refactoring

olegivchenkolula committed:
Code refactoring

olegivchenkolula committed:
Fix HttpMethods.Get

olegivchenkolula committed:
Added logging

olegivchenkolula committed:
Added logging

olegivchenkolula committed:
Merge from LulaSafe

xperiandri commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -17,9 +18,11 @@ type private GQLRequestContent =
       OperationName : string voption
       Variables : ImmutableDictionary<string, JsonElement> voption }
-
+      
```
Use https://marketplace.visualstudio.com/items?itemName=MadsKristensen.TrailingWhitespaceVisualizer to avoid this

xperiandri commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -9,6 +9,7 @@ open Giraffe
 open Microsoft.AspNetCore.Http
 open FSharp.Data.GraphQL
 open FSharp.Data.GraphQL.Types
+open Microsoft.Extensions.Logging
```
order

xperiandri commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -37,22 +40,25 @@ module HttpHandlers =
             let toResponse documentId =
                 function
                 | Direct   (data, errs) ->
+                    logger.LogInformation($"GraphQL toResponse Direct with documentId = '{documentId}', Data = {JsonSerializer.Serialize(data, Json.serializerOptions)}")
```
You need to use structural logging here
```suggestion
                    logger.LogInformation("Produced direct GraphQL response with documentId = '{documentId}' and data = {data}", documented, data)
```
As I understand how it works, it will do log record creation only if log level is Information or above. So we must not do any heavy duty logic there. It will happen under the hood

xperiandri commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -76,7 +82,7 @@ module HttpHandlers =
             then
                 let! result = Schema.executor.AsyncExecute (Introspection.IntrospectionQuery)
-                printfn "Result metadata: %A" result.Metadata
+                logger.LogInformation($"GraphQL Result metadata: '{result.Metadata}'")
```
```suggestion
                logger.LogInformation($"Produced GraphQL introspection result with documentId = '{documentId}' and data = {data} and metadata = {metadata}", documented, data, metadata)
```

xperiandri commented on samples/star-wars-api/Startup.fs:
```diff
@@ -39,4 +39,6 @@ type Startup private () =
             //.UseMiddleware<GraphQLWebSocketMiddleware<Root>>(Schema.executor, fun () -> { RequestId = Guid.NewGuid().ToString() })
             .UseGiraffe HttpHandlers.webApp
 
+            
+
```
```suggestion
```

olegivchenkolula committed:
Passed OperationName to AsyncExecute, refactored code

olegivchenkolula committed:
Passed OperationName to AsyncExecute, refactored code

xperiandri commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -37,22 +39,25 @@ module HttpHandlers =
                 | Stream data ->
-                    // TODO: Print to logger
+                    logger.LogInformation("Produced stream GraphQL response with documentId = '{documentId}' and data = {data}", documentId, data)
                     data |> Observable.add (fun d -> printfn "Subscription data: %s" (JsonSerializer.Serialize(d, Json.serializerOptions)))
```
And here to log too

xperiandri commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -37,22 +39,25 @@ module HttpHandlers =
                 | Deferred (data, errs, deferred) ->
-                    // TODO: Print to logger
+                    logger.LogInformation("Produced deferred GraphQL response with documentId = '{documentId}' and data = {data}", documentId, data)
                     deferred |> Observable.add (fun d -> printfn "Deferred: %s" (JsonSerializer.Serialize(d, Json.serializerOptions)))
```
And here to log too

xperiandri commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -87,20 +94,19 @@ module HttpHandlers =
-                        let! result = Schema.executor.AsyncExecute (query)
-                        printfn "Result metadata: %A" result.Metadata
+                        let! result = Schema.executor.AsyncExecute (query, operationName = request.OperationName.Value)
+                        logger.LogInformation($"GraphQL Result metadata: '{result.Metadata}'")
                         let response = toResponse result.DocumentId result.Content
```
Do we have something interesting to log after having result or response?

olegivchenkolula committed:
Code refactoring

olegivchenkolula committed:
Code refactoring

xperiandri commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -38,22 +42,28 @@ module HttpHandlers =
-                    // TODO: Print to logger
-                    deferred |> Observable.add (fun d -> printfn "Deferred: %s" (JsonSerializer.Serialize(d, Json.serializerOptions)))
+                    if not <| logger.IsEnabled LogLevel.Trace then
+                        logger.LogTrace("Produced deferred GraphQL response with documentId = '{documentId}' and data = {data} and deferred = {deferred}", documentId, data, deferred)
```
`deferred` is `IObservable` as I remember. We don't need it as nothing interesting there

xperiandri commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -76,34 +86,27 @@ module HttpHandlers =
-                        let response = toResponse result.DocumentId result.Content
-                        return Results.Ok response
-                    }
+            logger.LogTrace($"GraphQL query:{Environment.NewLine}{{query}}", query)
```
```suggestion
            logger.LogTrace($"Executing GraphQL query:{Environment.NewLine}{{query}}", query)
```

xperiandri commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -76,34 +86,27 @@ module HttpHandlers =
             if (request.Method = HttpMethods.Get || not <| hasData ())
             then
                 let! result = Schema.executor.AsyncExecute (Introspection.IntrospectionQuery)
```
```suggestion
                logger.LogInformation("Executing GraphQL introspection query)
                let! result = Schema.executor.AsyncExecute (Introspection.IntrospectionQuery)
```

xperiandri commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -76,34 +86,27 @@ module HttpHandlers =
             then
                 let! result = Schema.executor.AsyncExecute (Introspection.IntrospectionQuery)
-                printfn "Result metadata: %A" result.Metadata
+                logger.LogTrace("Producing GraphQL introspection result with documentId = '{documentId}' and data = {data} and metadata = {metadata}", result.DocumentId, result.Content, result.Metadata)
```
```suggestion
                logger.LogInformation($"Produced GraphQL result with documentId = '{{documentId}}' and metadata:{Environment.NewLine}{{metadata}}", result.DocumentId, result.Metadata)
                if logger.IsEnabled LogLevel.Trace then
                    logger.LogTrace($"GraphQL result data:{Environment.NewLine}{{data}}", result.Content)

```

xperiandri commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -38,22 +42,28 @@ module HttpHandlers =
                 | Deferred (data, errs, deferred) ->
-                    // TODO: Print to logger
-                    deferred |> Observable.add (fun d -> printfn "Deferred: %s" (JsonSerializer.Serialize(d, Json.serializerOptions)))
+                    if not <| logger.IsEnabled LogLevel.Trace then
```
```suggestion
                    if logger.IsEnabled LogLevel.Trace then
```

xperiandri commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -38,22 +42,28 @@ module HttpHandlers =
-                    // TODO: Print to logger
-                    data |> Observable.add (fun d -> printfn "Subscription data: %s" (JsonSerializer.Serialize(d, Json.serializerOptions)))
+                | Stream stream ->
+                    if not <| logger.IsEnabled LogLevel.Trace then
```
```suggestion
                    if logger.IsEnabled LogLevel.Trace then
```

xperiandri commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -76,34 +86,27 @@ module HttpHandlers =
+            let root = { RequestId = System.Guid.NewGuid().ToString () }
+            let query = removeWhitespacesAndLineBreaks query
+            let! result = Schema.executor.AsyncExecute (query, root, ?variables = variables, ?operationName = operationName)
+            logger.LogTrace("GraphQL Result metadata: '{metadata}'", result.Metadata)
```
```suggestion
```

xperiandri commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -38,22 +42,28 @@ module HttpHandlers =
+            let logger = ctx.RequestServices.CreateLogger moduleType
+
             // TODO: validate the result
             let toResponse documentId =
                 function
```
```suggestion
            let toResponse result =
                match result.Content
```
so that we can output metadata too as in code above

xperiandri commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -38,22 +42,28 @@ module HttpHandlers =
             let toResponse documentId =
                 function
                 | Direct   (data, errs) ->
+                    if not <| logger.IsEnabled LogLevel.Trace then
+                        logger.LogTrace("Produced direct GraphQL response with documentId = '{documentId}' and data = {data}", documentId, JsonSerializer.Serialize(data, Json.serializerOptions))
```
```suggestion
                    logger.LogInformation($"Produced direct GraphQL response with documentId = '{{documentId}}' and metadata:{Environment.NewLine}{{metadata}}", documentId, metadata)
                    if not <| logger.IsEnabled LogLevel.Trace then
                        logger.LogTrace($"GraphQL response data:{Environment.NewLine}:{{data}}", JsonSerializer.Serialize(data, Json.serializerOptions))
```

olegivchenkolula committed:
Code refactoring

olegivchenkolula committed:
Code refactoring

Commit:
Removed version from `global.global.json`

xperiandri commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -1,25 +1,29 @@
 open FSharp.Data.GraphQL
 open FSharp.Data.GraphQL.Types
+open Giraffe
+open System.Text.Json.Serialization
```
Order

xperiandri commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -38,22 +42,28 @@ module HttpHandlers =
+            let logger = ctx.RequestServices.CreateLogger moduleType
+
             // TODO: validate the result
             let toResponse documentId =
```
Let's copy my new implementation from Risk.
Try to execute scenarios that are possible (you can try to test them using React samples).

xperiandri commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -75,35 +85,29 @@ module HttpHandlers =
             else
 
-            let! request = ctx.BindJsonAsync<GQLRequestContent>()
+            //let! request = ctx.BindJsonAsync<GQLRequestContent>()
+            let! request = JsonSerializer.DeserializeAsync<GQLRequestContent>(ctx.Request.Body, Json.serializerOptions)
```
With `Skippable` this have to work now

xperiandri commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -75,35 +85,29 @@ module HttpHandlers =
+
+            let root = { RequestId = System.Guid.NewGuid().ToString () }
+            let query = removeWhitespacesAndLineBreaks query
+            let! result = Schema.executor.AsyncExecute (query, root, ?variables = variables, ?operationName = operationName)
+            logger.LogTrace("GraphQL Result metadata: '{metadata}'", result.Metadata)
+            let response = toResponse result.DocumentId result.Content
+            return Results.Ok response
```
Update as in Risk

xperiandri reviewed (changes requested):
@olegivchenkolula use the latest changes from Risk

olegivchenkolula committed:
Merge branch 'LHK-551-Change-printing-from-console-to-log' of https://github.com/Lula-Technologies-Inc/FSharp.Data.GraphQL into LHK-551-Change-printing-from-console-to-log

olegivchenkolula committed:
Code refactoring, changed logs,

andrii-berestov-lula commented on global.json:
```diff
@@ -1,6 +1,5 @@
 {
   "sdk": {
-    "version": "6.0.400",
```
@olegivchenkolula if we delete version from `sdk` it will have an error during build

xperiandri replied:
Strange, works on my device

andrii-berestov-lula commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -37,23 +43,58 @@ module HttpHandlers =
 
     let private graphQL (next : HttpFunc) (ctx : HttpContext) =
         task {
+            let logger = ctx.RequestServices.CreateLogger moduleType
```
We can inject logger from DI, like here
https://github.com/giraffe-fsharp/Giraffe/issues/10

awbv replied:
it already uses `ILoggerFactory`, so it's fine

andrii-berestov-lula committed:
[LHK-551] fix launchSettings.json to run /voyager, instead of //voyager

xperiandri commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -62,48 +103,47 @@ module HttpHandlers =
+                let! bytesRead = body.ReadAsync(buffer, 0, 1)
+                body.Seek(0, SeekOrigin.Begin) |> ignore
+                return (bytesRead > 0)
+            }
```
```suggestion
            let hasData () = task {
	            if request.Body.CanSeek then
	                return (request.Body.Length > 0L)
	            else
	                request.EnableBuffering()
	                let body = request.Body
	                let buffer = Array.zeroCreate 1
	                let! bytesRead = body.ReadAsync(buffer, 0, 1)
	                body.Seek(0, SeekOrigin.Begin) |> ignore
	                return (bytesRead > 0)
            }

```

xperiandri commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -62,48 +103,47 @@ module HttpHandlers =
                 return Results.Ok response
             else
-
+            let! hd = hasData()
+            if (not hd)
+            then
+                logger.LogInformation("Executing GraphQL introspection query")
+                let! result = Schema.executor.AsyncExecute (Introspection.IntrospectionQuery)
+                let response = result |> toResponse
+                return Results.Ok response
+            else
```
Let's modify this a bit.
As we discovered that introspection query can come in a body too, we can add `checkIsIntrospectionQuery` function and then if on it.
``` F#
[<Struct>]
type OperationType =
	| IntrospectionQuery of string option
	| OperationQuery of GQLRequestContent

let checkIsNotIntrospectionQuery = task {
    if isGet then IntrospectionQuery ValueNone
    else
    let! hasbody = checkIfHasBody()
    if not hasBody then IntrospectionQuery ValueNone
    else
    let! request = ctx.BindJsonAsync<GQLRequestContent>()
    // This needs to be verified and can be omitted in the current PR
    if Introspection.IntrospectionQuery.Contains request.Query then IntrospectionQuery request.Query
    else OperationQuery request
}

let executeIntrospectionQuery query = task {
	let! result =
		match query with
		| ValueNone ->
			logger.LogInformation("Executing default GraphQL introspection query")
			Schema.executor.AsyncExecute (Introspection.IntrospectionQuery)
		| ValueSome query ->
			logger.LogInformation($"Executing GraphQL introspection query:{Environment.NewLine}{query}", query)
			Schema.executor.AsyncExecute (query)
		let response = result |> toResponse
		return Results.Ok response
}

let executeOperation request = task {
	...
}

match! checkIsNotIntrospectionQuery with
| IntrospectionQuery query -> return! executeIntrospectionQuery query 
| OperationQuery request -> return! executeOperation request
```

andrii-berestov-lula committed:
[LHK-551] code review changes

xperiandri commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -37,74 +48,129 @@ module HttpHandlers =
+                    if not hasBody then return IntrospectionQuery ValueNone
+                    else
+                        let! request = ctx.BindJsonAsync<GQLRequestContent>()
+                        if Introspection.IntrospectionQuery.Contains request.Query
```
Waiting for the experimental results of this line

GBirkel committed:
Minor code comment.

GBirkel committed:
Minor function rename

GBirkel committed:
Merge branch 'refs/heads/LulaSafe' into LHK-551-Change-printing-from-console-to-log
a

GBirkel committed:
Build now succeeds but introspection test still fails. Some TODO notes.

GBirkel committed:
Looks like there is a redundant copy of NameValueLookup here.

GBirkel committed:
Swapping two variable names for consistency.

GBirkel committed:
Some code comments

GBirkel committed:
Minor changes

GBirkel committed:
Marging in changes, need to manually fix samples/star-wars-api/Properties/launchSettings.json
Merge branch 'refs/heads/LulaSafe' into LHK-551-Change-printing-from-console-to-log

GBirkel committed:
Restoring samples/star-wars-api/Properties/launchSettings.json from LulaSafe

GBirkel committed:
Looks like this needs to be removed for a good build

GBirkel committed:
Don't need this

xperiandri commented on samples/star-wars-api/HttpHandlers.fs:
```diff
@@ -33,68 +44,132 @@ module HttpHandlers =
 
-            let removeWhitespacesAndLineBreaks (str : string) = str.Trim().Replace ("\r\n", " ")
+                | RequestError errs ->
+                    GQLResponse.RequestError (documentId, errs)
```
I will leave this comment not to forget about
@GBirkel

xperiandri commented on tests/FSharp.Data.GraphQL.Tests/VariablesTests.fs:
```diff
@@ -45,10 +45,10 @@ let TestInputObject =
         Define.Input("c", StringType)
         Define.Input("d", Nullable TestComplexScalar)
-        Define.Input("e", Nullable( InputArrayOf (Nullable StringType)) )
+        Define.Input("e", Nullable( InputArrayOf (Nullable StringType)))
```
And about formatting with Fantomas here
@GBirkel
